### PR TITLE
LaTeX rendering in element-web using KaTeX

### PR DESCRIPTION
**Pull request by akissinger**
This is an implementation of MSC2191: matrix-org/matrix-doc#2191

It is based on a trimmed-down and updated version of PR matrix-org/matrix-react-sdk#3251, which won't auto-merge any more. It is different in that implements the latest edits (as of Sept 2020) of MSC2191 and does not include GUI configuration (only a global on/off switch in Labs and flags in for custom delimiters in `config.json`).

LaTex rendering is implemented using KaTeX 0.12.0, and is turned off by default. It can be switched on by the user in Labs (documentation updated in matched PR vector-im/element-web#15277).

The other relevant setting is `latex_maths_delims`, which allows customization of user-facing LaTeX delimiters. By default, inline-math is wrapped in double-$'s and display in triple-$'s. The inline case matches the conventions of other chat utilities with TeX support, such as Zulip.

You can reproduce the default behaviour in `config.json` as follows:
```
"latex_maths_delimiters": {
  "inline_left": "$$",
  "inline_right": "$$",
  "inline_pattern": "\\$\\$(([^$]|\\\\\\$)*)\\$\\$",
  "display_left": "$$$",
  "display_right": "$$$",
  "display_pattern": "\\$\\$\\$(([^$]|\\\\\\$)*)\\$\\$\\$"
}
```
The `*_pattern` settings are given explicitly, which are double-quoted strings that get passed to `RegExp(..., "mg")` internally. Note backslashes in the regex should be escaped, and literal backslashes in the regex need to be escaped twice.

The first match group is HTML-entities encoded and given as the string in `<span data-mx-maths="....">` or `<div data-mx-maths="....">`, for inline and display math, respectively.

The translation to/from delimited LaTeX to HTML tags described in MSC2191 is done in `src/editor/(de)serialize.ts`. The KaTeX rendering is in `src/HtmlUtils.tsx`.

Finally, `src/Markdown.js` is told to ignore these tags, so clients can mix math and markdown in a single message.

`Signed-off-by: Aleks Kissinger <aleks0@gmail.com>`

### CODE REVIEW

**Comment by NickHu** [comment]
Might be pertinent to add new config stuff to `config.sample.json` too

**Comment by akissinger** [comment]
I can't seem to include `katex.min.css` in a way that `jest-tests` agrees with. I tried putting an `@import` in `_common.scss`, but it won't pull the fonts correctly then.

**Comment by akissinger** [comment]
Shifted katex.min.css into webpack.config.json, which makes most of the tests happy. Don't know why the first 2 are failing and/or if it's my doing.

**Comment by behrmann** [comment]
@akissinger You can bisect wehre the problem is by rolling back your branch to an earlier version. If I remember correctly all tests passed when you first submitted this branch.

Looking at the failing tests, a node module called dom_handler seems to be broken. No idea how this relates to your work. How does it look, when you build locally?

**Comment by behrmann** [comment]
Looks like d2054ea is good and build errors where introduced later.

**Comment by akissinger** [comment]
Tidied the commits and works now. Actual code is identical to before, so I figure it was a one-off problem with the CI server.

**Comment by behrmann** [comment]
Excellent, I hope your work gets reviewed and merged soon.

**Review by beinganukul — APPROVED** [review]

File: src/editor/serialize.ts
Comment (uhoreg) [inline src/editor/serialize.ts:56]: Since the maths processing is done before the markdown processing, there is the potential for some odd interactions.  If the LaTeX happens to have valid markdown, then the fallback will have the markdown applied.  For example, if someone types `$$ *x* $$`, then it will produce `<span data-mx-maths=\" *x* \"><code> <em>x</em> </code></span>` (note the `<em>x</em>` in the fallback).

It seems like this type of thing will be rare, and fortunately neither `$$ x*y*z $$` nor `$$ x * y * z $$` seem to trigger this.  But it would be nice if we could avoid this if possible.  It may be necessary to fiddle with the `toHtml` function in `src/Markdown.js`, and maybe see if you can coerce commonmark to output the literal contents of `<code>` blocks inside of data-mx-maths tags.

File: src/Markdown.js
Comment (uhoreg) [inline src/Markdown.js:42]: As discussed in the room, there is *something* that is causing the parser to fail to catch just a plain `$$$x$$$` is math.  If there is (non-whitespace) text before (e.g. `foo $$$x$$$`), then it works fine.  I suspect that it is something in the `isPlainText` function in this file, possibly around the `if (TEXT_NODES.indexOf(node.type) > -1)` clause that causes it to think that there's no markup, so it just emits the plain text version.

**Review by uhoreg — CHANGES_REQUESTED** [review]
Thanks for your work on this!  It's a great start.

File: src/Markdown.js
Comment (turt2live) [reply src/Markdown.js:21]: let's limit it to just the math blocks - people can already use backticks as needed.

**Comment by uhoreg** [comment]
The test failure (chains-end-to-end-tests) is probably just because you're based on an older version of `develop`.  If you merge the latest `develop` it should pass.

File: src/Markdown.js
Comment (uhoreg) [reply src/Markdown.js:42]: Something similar happens when you enter `<div> **foo** </div>` -- it results in the message being sent as text without any markdown processing (the `foo` is not bolded).  That indicates that the bug isn't with your code, but math markup makes it more obvious since nobody is going to type `<div> **foo** </div>`.  So it probably makes sense to fix it as part of this PR (assuming they're caused by the same thing).

File: src/HtmlUtils.tsx
Comment (akissinger) [reply src/HtmlUtils.tsx]: fixed in 65c4460ab

File: src/Markdown.js
Comment (akissinger) [reply src/Markdown.js]: fixed in 919a1a812

File: src/Markdown.js
Comment (akissinger) [reply src/Markdown.js]: also fixed in 919a1a812

File: src/HtmlUtils.tsx
Comment (akissinger) [reply src/HtmlUtils.tsx]: Done in 96742fc. i18n strings for EN and NL in a89adb8. Documentation in the matching PR vector-im/element-web#15277.

Should the delimiters be done this way too, or is SdkConfig fine? It seems more sensible that this should be server-level (or maybe just not have it as a setting at all).

Comment (uhoreg) [reply src/HtmlUtils.tsx]: I think it's fine to leave the delimiter settings as SdkConfig for now.

File: src/Markdown.js
Comment (akissinger) [reply src/Markdown.js]: this change is no longer needed, since code blocks are being added after markdown parsing in cc713af (see below)

File: src/editor/serialize.ts
Comment (akissinger) [reply src/editor/serialize.ts:58]: after trying all sorts of stuff to get commonmark to ignore bits of text, i found the much simpler solution was to just add the `<code>` block with tex in after markdown processing (cc713af). Now the only change needed in Markdown.js is allowing the math tags through.

File: src/HtmlUtils.tsx
Comment (akissinger) [reply src/HtmlUtils.tsx]: fixed 10b73213

File: src/HtmlUtils.tsx
Comment (uhoreg) [inline src/HtmlUtils.tsx:33]: Although `cheerio` is a dependency via some other dependency, it isn't specified in our `package.json`, which means that if its reverse-dependency ever stops depending on it, this will break.  So `cheerio` should be added to our `package.json` directly.

File: src/editor/serialize.ts
Comment (uhoreg) [reply src/editor/serialize.ts:58]: That seems like a nice solution.

File: src/HtmlUtils.tsx
Comment (akissinger) [reply src/HtmlUtils.tsx:33]: done in 173d798

File: src/Markdown.js
Comment (akissinger) [reply src/Markdown.js:41]: This is now fixed for the case of latex math in 2204e6c. The case you pointed out actually doesn't see to be a bug, but the expected (but weird) behaviour of HTML blocks, according to the Commonmark spec: https://spec.commonmark.org/0.29/#html-blocks. This was pointed out by @rda in #pigeon.digital:matrix.org.

**Comment by akissinger** [comment]
okay, that's all of @uhoreg's changes done. For some reason the local jest tests don't like calling `Settings.getValue` inside of a unit test. I don't know if the problem is with me or the test setup. Otherwise, everything else is passing.

**Comment by behrmann** [comment]
@akissinger I just poked a bit through the code and I think the problem might be with the test. If I understand it correctly `test/components/views/messages/TextualBody-test.js` mocks out the `MatrixClient` object behind `MatrixClientPeg.matrixClient` with some stubs, leaving out `isGuest` in all of them, which should mean that it's null. Since `readFeature` uses this particular attribute and tries to call it, the test breaks. I guess it if you add a lambda for `isGuest` to the stubs in the test, it should work.

Take this analysis with a rather large portion of salt, though, I do neither JS nor TS work and this is just a non-educated guess stemming from the wish to see this PR land soon.

**Comment by akissinger** [comment]
Added stubs for `isGuest()` to `TextualBody-test.js`, and the tests all pass now. Thanks @behrmann!

File: src/editor/serialize.ts
Comment (NickHu) [inline src/editor/serialize.ts:51]: Seeing as both fluffy chat (https://gitlab.com/famedly/famedlysdk/-/merge_requests/501) and the Element-Android PR (https://github.com/vector-im/element-android/pull/2133) adhere to "standard" LaTeX conventions rather than Zulip's, it might be better to default to that instead. I've had to remind people to use double `$$` very frequently for inline maths so far because they are so used to `$x + y$`, as opposed to Zulip conventions.

Comment (akissinger) [reply src/editor/serialize.ts:49]: done in 839bae2. Since this is now a "labs" feature that users can switch off and on, i think it's reasonable to set it to single $'s by default

File: package.json
Comment (uhoreg) [inline package.json:80]: Note to the next reviewer: cheerio is already pulled in by another dependency, so this isn't adding a new dependency, but is added here because it is now being used by `HtmlUtils`

File: src/i18n/strings/en_EN.json
Comment (uhoreg) [inline src/i18n/strings/en_EN.json:702]: We use UK English for the untranslated string, which I believe should be "LaTeX maths ..."  here and in the other translation files, and in `settings/Settings.ts`  (The en_US translation would be "LaTeX math ...")

**Review by uhoreg — APPROVED** [review]
Aside from a translation nit, looks good to me.  Seems to do what it says on the box.

This should be re-reviewed by another developer who's more familiar with the react-sdk.  It will probably need design/product review before the labs flag is removed.

Thank you for working on this, and sorry for the delay in reviewing it.

File: src/i18n/strings/nl.json
Comment (turt2live) [inline src/i18n/strings/nl.json:202]: translating like this causes conflicts - please only use the output of `yarn i18n` and don't populate other json files.

**Review by turt2live — CHANGES_REQUESTED** [review]
The code seems reasonable at first glance, though I'm going to leave it to someone who has a bit more experience with the composer.

File: src/i18n/strings/nl.json
Comment (akissinger) [reply src/i18n/strings/nl.json]: reverted. I don't fully understand how i18n works, so I'll leave this to someone else.

File: src/HtmlUtils.tsx
Comment (jryans) [inline src/HtmlUtils.tsx:421]: Let's move this inside the settings block, so we can skip this work if the feature is not enabled.

File: src/i18n/strings/en_US.json
Comment (jryans) [inline src/i18n/strings/en_US.json:131]: Please revert this translation here, as it can cause conflicts. Once the main PR has merged, you can submit this translation via https://translate.element.io.

File: src/settings/Settings.ts
Comment (jryans) [inline src/settings/Settings.ts:121]: Personally I would suggest making it more active with a verb, and avoiding math vs. maths by just removing the word:

```suggestion
        displayName: _td("Render LaTeX in messages"),
```

If you'd like to keep "maths" in there, then please use UK spelling.

**Review by jryans — CHANGES_REQUESTED** [review]
Overall, this looks great, and I am excited to try it out! 😄

I believe there are only a few small nits to resolve, and then we should be ready to merge this.

File: src/settings/Settings.ts
Comment (akissinger) [reply src/settings/Settings.ts]: I decided to go with "Render LaTeX maths in messages". I think saying "maths" here is useful, since all tex formatting is restricted to between the dollar-signs, and general text formatting should (still) be done with markdown.

**Comment by akissinger** [comment]
@jryans okay, I think that's all your changes done. I don't know why the new Vercel thing isn't working, but it might just need to be re-merged to develop. in any case, the other tests still pass.

**Comment by jryans** [comment]
> I don't know why the new Vercel thing isn't working, but it might just need to be re-merged to develop. in any case, the other tests still pass.

Ah yeah, sorry about this, I was trialling Vercel to add previews of PR, but it went a bit haywire. I have disabled it now, and the error is fine to ignore.

**Review by jryans — APPROVED** [review]
Looks great, thanks again for working on this! 😄

**Comment by akissinger** [comment]
nice! it wasn't exactly the one-liner I was expecting, but glad to see it merged. :)

**Comment by dinojr** [comment]
> Lab



> LaTex rendering is implemented using KaTeX 0.12.0, and is turned off by default. It can be switched on by the user in Labs

I'm not familiar with this "Labs" configuration since I'm only using the matrix.org server. I'm guessing it's where the server is configured. As for the matrix.org server, do they plan to enable LaTeX rendering ?

**Comment by t3chguy** [comment]
It is not related to the Matrix server you are using but instead the client you are using. develop.element.io has a config.json with Labs enabled.

**Comment by dinojr** [comment]
>  https://develop.element.io has a config.json with Labs enabled.

I'm using element on Android. There is indeed a Labs settings but so far without an option to enable LaTeX. I guess it should appear there in a later version (I'm using version 1.0.11[401500112], up to date in the beta program from the Android PlayStore).

However, I can't find a Labs tab/menu in the desktop client for Linux (from https://element.io/get-started).

However, the

**Comment by t3chguy** [comment]
This repository is not related to the Mobile apps. The mobile apps are native, this is a React SDK.

For enabling Labs on the Desktop app see https://github.com/vector-im/element-web/blob/develop/docs/config.md and https://github.com/vector-im/element-desktop#user-specified-configjson

**Comment by dinojr** [comment]
> For enabling Labs on the Desktop app see https://github.com/vector-im/element-web/blob/develop/docs/config.md and https://github.com/vector-im/element-desktop#user-specified-configjson

Thanks a lot, LaTeX is now working for me.

**Comment by NickHu** [comment]
For what it's worth, I implemented support for this in Element Android here: https://github.com/vector-im/element-android/pull/2133

It could do with some polish but it basically works.

**Comment by LeonardKoenig** [comment]
> This repository is not related to the Mobile apps. The mobile apps are native, this is a React SDK.
> 
> For enabling Labs on the Desktop app see https://github.com/vector-im/element-web/blob/develop/docs/config.md and https://github.com/vector-im/element-desktop#user-specified-configjson

Hm, I couldn't find any config.json in the packaged version of my system, but you can do:

1. go to a room
2. upper right go to the (i) symbol
3. room settings cog wheel
4. advanced
5. open devtools
6. settings explorer
7. change `feature_latex_maths` to:
    ```
   {
        "device": true,
        "room-device": null,
        "room-account": null,
        "account": null,
        "room": null,
        "config": null,
        "default": false
    }
    ```

**Pull request merged**
